casper/src/main/scala/coop/rchain/casper/MultiParentCasper.scala

```diff
@@ -59,182 +55,205 @@ object MultiParentCasper {
+
+      // Calculate finalized fringe from justifications
+      msgMap    = dag.dagMessageState.msgMap
+      parents   = parentHashes.map(msgMap)
+      finalizer = Finalizer(dag.dagMessageState.msgMap)
+      // Get currently finalized bonds map
+      prevFringe       = finalizer.latestFringe(parents)
+      prevFringeHashes = prevFringe.map(_.id)
+      // Previous fringe state should be present (loaded from BlockMetadata store)
+      fringeRecord <- dag.fringeStates
+                       .get(prevFringeHashes)
+                       .liftTo {
+                         val fringeStr = PrettyPrinter.buildString(prevFringeHashes)
+                         val errMsg =
+                           s"Fringe state not available in state cache, fringe: $fringeStr"
+                         new Exception(errMsg)
+                       }
+
+      (prevFringeState, prevFringeRejectedDeploys) = fringeRecord
+      prevFringeStateHash                          = prevFringeState.bytes.toArray.toByteString
+      // TODO: for empty fringe bonds map should be loaded from bonds file (if validated in replay)
+      bondsMap <- if (prevFringe.isEmpty)
+                   justifications.head.bondsMap.pure[F]
+                 else
+                   RuntimeManager[F].computeBonds(prevFringeStateHash)
+
+      (_, newFringeOpt) = finalizer.calculateFinalization(parents, bondsMap)
+      newFringeHashes   = newFringeOpt.map(_.map(_.id))
+
+      prevFringeStr      = PrettyPrinter.buildString(prevFringeHashes)
+      prevFringeStateStr = PrettyPrinter.buildString(prevFringeState.toByteString)
+      _                  = println(s"PREV FRINGE $prevFringeStr")
+      _                  = println(s"PREV FRINGE STATE $prevFringeStateStr")
+
+      // If new fringe is finalized, merge it
+      newFringeResult <- newFringeHashes.traverse { fringe =>
+                          val seenByPrevFringe = prevFringeHashes.map(msgMap).flatMap(_.seen)
+                          val seenByFringe     = fringe.map(msgMap).flatMap(_.seen)
+                          val fringeBlocks     = seenByFringe -- seenByPrevFringe
+                          for {
+                            // Get all blocks in the fringe to be merged
+                            result <- DagMerger.mergeFringe[F](
+                                       fringeBlocks,
+                                       prevFringeState,
+                                       InterpreterUtil.getBlockIndex(_).map(_.deployChains),
+                                       RuntimeManager[F].getHistoryRepo,
+                                       DagMerger.costOptimalRejectionAlg
+                                     )
+
+                            fringeBlocksCount = fringeBlocks.size
+                            fringeDiffBlocks  = PrettyPrinter.buildString(fringeBlocks)
+                            _                 = println(s"FRINGE DIFF ($fringeBlocksCount) $fringeDiffBlocks")
```

Comment:
`Log` instead of `println`

Reply:
It's left over of debugging, I'll remove it.